mirgecom/navierstokes.py

```diff
@@ -0,0 +1,416 @@
+
+    Returns
+    -------
+    :class:`numpy.ndarray`
+
+        Array of :class:`~mirgecom.fluid.ConservedVars` representing the
+        gradient of the fluid conserved variables.
+    """
+    dd_vol_quad = DOFDesc("vol", quadrature_tag)
+    dd_faces_quad = DOFDesc("all_faces", quadrature_tag)
+
+    if operator_states_quad is None:
+        operator_states_quad = make_operator_fluid_states(
+            discr, state, gas_model, boundaries, quadrature_tag)
+
+    vol_state_quad, inter_elem_bnd_states_quad, domain_bnd_states_quad = \
+        operator_states_quad
+
+    get_interior_flux = partial(
+        _gradient_flux_interior, discr, numerical_flux_func)
+
+    cv_interior_pairs = [TracePair(state_pair.dd,
+                                   interior=state_pair.int.cv,
+                                   exterior=state_pair.ext.cv)
+                         for state_pair in inter_elem_bnd_states_quad]
+
+    cv_flux_bnd = (
+
+        # Domain boundaries
+        sum(bdry.cv_gradient_flux(
+            discr,
+            # Make sure we get the state on the quadrature grid
+            # restricted to the tag *btag*
+            as_dofdesc(btag).with_discr_tag(quadrature_tag),
+            gas_model=gas_model,
+            state_minus=domain_bnd_states_quad[btag],
+            time=time,
+            numerical_flux_func=numerical_flux_func)
+            for btag, bdry in boundaries.items())
+
+        # Interior boundaries
+        + sum(get_interior_flux(tpair) for tpair in cv_interior_pairs)
+    )
+
+    # [Bassi_1997]_ eqn 15 (s = grad_q)
+    return grad_operator(
+        discr, dd_vol_quad, dd_faces_quad, vol_state_quad.cv, cv_flux_bnd)
+
+
+def grad_t_operator(
+        discr, gas_model, boundaries, state, *, time=0.0,
+        numerical_flux_func=gradient_flux_central,
+        quadrature_tag=DISCR_TAG_BASE,
+        # Added to avoid repeated computation
+        # FIXME: See if there's a better way to do this
+        operator_states_quad=None):
+    r"""Compute the gradient of the fluid temperature.
+
+    Parameters
+    ----------
+    state: :class:`~mirgecom.gas_model.FluidState`
+
+        Fluid state object with the conserved state, and dependent
+        quantities.
+
+    boundaries
+        Dictionary of boundary functions keyed by btags
+
+    time
+        Time
+
+    gas_model: :class:`~mirgecom.gas_model.GasModel`
+
+        Physical gas model including equation of state, transport,
+        and kinetic properties as required by fluid state
+
+    quadrature_tag
+        An identifier denoting a particular quadrature discretization to use during
+        operator evaluations.
+
+    Returns
+    -------
+    :class:`numpy.ndarray`
+
+        Array of :class:`~meshmode.dof_array.DOFArray` representing the gradient of
+        the fluid temperature.
+    """
+    dd_vol_quad = DOFDesc("vol", quadrature_tag)
+    dd_faces_quad = DOFDesc("all_faces", quadrature_tag)
+
+    if operator_states_quad is None:
```

Comment:
Is there a situation where this *wouldn't* be passed?

Reply:
Yes.  In almost all situations, NS computes its own operator states.  The exception is those cases where the operator states have already been computed by another component (like in wall model runs).  This was part of a refactor to support the wall model.

Reply:
And for gradients: visualization